core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -157,22 +165,23 @@ public int getOnReadyThreshold() {
    */
   public void setOnReadyThreshold(int onReadyThreshold) {
     checkArgument(onReadyThreshold > 0, "onReadyThreshold must be > 0");
-    boolean doNotify;
+    boolean doNotify = false;
     synchronized (onReadyLock) {
-      if (this.onReadyThreshold <= numSentBytesQueued && onReadyThreshold > numSentBytesQueued) {
-        shouldNotifyOnReady = true;
-      }
+      doNotify = wentBelowThreshold(
+          numSentBytesQueued,
+          this.onReadyThreshold, // old threshold
+          numSentBytesQueued,
+          onReadyThreshold); // new threshold.
       this.onReadyThreshold = onReadyThreshold;
-      doNotify = needToNotifyOnReady();
```

Comment:
Oh I guess we still are ... but we should just call `listener().onReady()` directly, so that we don't have to lock again. WDYT?